File: leetcode.com/1008_construct_binary_search_tree_from_preorder_traversal/solution.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

inline void insert(struct TreeNode *root, int value)
{
    struct TreeNode *node = malloc(sizeof(struct TreeNode));

    node->val = value;
    node->left = NULL;
    node->right = NULL;

    while (root) {
        if (value > root->val) {
            if (root->right) {
                root = root->right;
            } else {
                root->right = node;
                break;
            }
        } else {
            if (root->left) {
                root = root->left;
            } else {
                root->left = node;
                break;
            }
        }
    }
}

struct TreeNode* bstFromPreorder(int* preorder, int preorderSize)
{
    struct TreeNode *node = malloc(sizeof(struct TreeNode));

    node->val = preorder[0];
    node->left = NULL;
    node->right = NULL;

    for (int i = 1 ; i < preorderSize ; i++)
        insert(node, preorder[i]);

    return node;
}

File: leetcode.com/1008_construct_binary_search_tree_from_preorder_traversal/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func bstAppend(node *TreeNode, val int) {
    for node != nil {
        if val >= node.Val {
            if node.Right == nil {
                node.Right = &TreeNode{val, nil, nil}
                return
            } else {
                node = node.Right
            }
        } else {
            if node.Left == nil {
                node.Left = &TreeNode{val, nil, nil}
                return
            } else {
                node = node.Left
            }
        }
    }
}

func bstFromPreorder(preorder []int) *TreeNode {
    root := &TreeNode{preorder[0], nil, nil}

    for i := 1 ; i < len(preorder) ; i++ {
        bstAppend(root, preorder[i])
    }

    return root
}
